runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/GreedyPCollectionFusers.java

```diff
@@ -0,0 +1,305 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.core.construction.graph;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import com.google.common.collect.ImmutableMap;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.beam.model.pipeline.v1.RunnerApi.Environment;
+import org.apache.beam.model.pipeline.v1.RunnerApi.PCollection;
+import org.apache.beam.model.pipeline.v1.RunnerApi.ParDoPayload;
+import org.apache.beam.runners.core.construction.PTransformTranslation;
+import org.apache.beam.runners.core.construction.graph.PipelineNode.PTransformNode;
+
+/**
+ * A Fuser that constructs a fused pipeline by fusing as many PCollections into a stage as possible.
+ */
+public class GreedyPCollectionFusers {
+  private static final Map<String, FusibilityChecker> URN_FUSIBILITY_CHECKERS =
+      ImmutableMap.<String, FusibilityChecker>builder()
+          .put(PTransformTranslation.PAR_DO_TRANSFORM_URN, GreedyPCollectionFusers::canFuseParDo)
+          .put(PTransformTranslation.WINDOW_TRANSFORM_URN, GreedyPCollectionFusers::canFuseWindow)
+          .put(PTransformTranslation.FLATTEN_TRANSFORM_URN, GreedyPCollectionFusers::canFuseFlatten)
+          .put(
+              PTransformTranslation.GROUP_BY_KEY_TRANSFORM_URN,
+              GreedyPCollectionFusers::canFuseGroupByKey)
+          .build();
+  private static final FusibilityChecker DEFAULT_FUSIBILITY_CHECKER =
+      GreedyPCollectionFusers::unknownTransformFusion;
+
+  // TODO: Migrate
+  private static final Map<String, CompatibilityChecker> URN_COMPATIBILITY_CHECKERS =
+      ImmutableMap.<String, CompatibilityChecker>builder()
+          .put(
+              PTransformTranslation.PAR_DO_TRANSFORM_URN,
+              GreedyPCollectionFusers::parDoCompatibility)
+          .put(
+              PTransformTranslation.WINDOW_TRANSFORM_URN,
```

Comment:
We should really call this ASSIGN_WINDOWS_TRANSFORM_URN or WINDOW_INTO_TRANSFORM_URN, WINDOW_TRANSFORM_URN is vague if you think about window mapping and window strategies.

Reply:
ACK